### ISSUE TITLE
#pragma once errors with simple first person UE5 project

### ISSUE THREAD

**Issue by max-cis** [boocs/ue4-intellisense-fixes#22]
Hi, 
I have used your extension to remove most of the problems but I am still getting a few. 
Unfortunately, I am required to use macos for this and I am not particularly familiar with various things. 

My current setup is an i7 iMac so not on the M1 or anything. I was hoping
I have set up the cppproperties for intelli sense to `"macos-clang-x64".` I left the compiler path to the default. Also, I am only using the option from your extension  `"UEIntellisenseFixes.enableFixes": true, "UEIntellisenseFixes.cppStandard":` "c++17"`. 

Any ideas on what I could do to fix this? I have attached a picture of the errors. Maybe it helps. 

Thanks,
M
<img width="1651" alt="Screenshot 2022-08-29 at 11 34 31" src="https://user-images.githubusercontent.com/47692294/187182858-ae6ab168-ebb4-430f-8098-5c33e553425e.png">

**Comment by boocs** [comment]
Some things to try:

Make sure you can build successfully first with the 'Editor' suffix config like so:
![image](https://user-images.githubusercontent.com/62588629/187189027-a3d016f4-7e7b-46f3-a4f8-245a3f8c0bb5.png)

This is a VSCode Build Task. TestFps is the project name and notice the 'Editor' suffix. Win64 might be Mac for you.

This will create some files needed for Intellisense if they're missing.

Now you'll need to reset your project which will add the new file paths to Intellisense configs.
![image](https://user-images.githubusercontent.com/62588629/187190600-c0a2af18-97e0-41e6-a06c-e318f8b8420a.png)

Right click on your *.uproject file and choose Generate Visual Studio Project Files (or something similiar). If you can't find it you can also do this by loading your project in to the UE5 Editor and choose this option from the Tools menu. It's called 'Refresh Visual Studio Code Project'. It's also possible inside VSCode. One of the debug configs is called Generate Project Files. You do need to install the Microsoft C# extension for it to work though.

Restart VSCode and see if it works.

Another thing to try, maybe the config you're using is bugged.
![image](https://user-images.githubusercontent.com/62588629/187191786-068deaa9-2cd7-448f-b49e-b11d2b8dc102.png)

You can click the Win32 which is the name of one of the intellisense configs. Yours will probably be named something different. When you click it you'll get a menu with two choices. Click the opposite config that's currently active.
![image](https://user-images.githubusercontent.com/62588629/187192271-c2a76ad0-1cbb-4bb8-86d0-a8f9233fab16.png)

**Comment by max-cis** [comment]
Thanks for getting back so quickly. 
I managed to figure out the reason behind the problems. 

I am not sure if it is because I didn't set up the compiler in your extension properly, but I needed to add the following two parameters to the `c_cpp_properties.json` file to ensure that all the errors went away. 
```  
  "includePath": ["${workspaceFolder}/**"],
  "compilerPath": "/usr/bin/clang",
```

I am not sure if you think it is worth adding this to the extension so it includes these two parameters when it corrects the file. 

I don't know how to do it, but I think could be marked as a workaround in GitHub. May help other MacOS users.

**Comment by boocs** [comment]
I do have a compiler path setting in my extension but the includePath making it work is weird.

It would mean that it's not using the compile commands file for Intellisense and instead using the cpptools includePath setting.

This could also mean you're not using the 'smart' Intellisense and instead using the other intellisense called 'Tag Parser'. 

Pretty strange. Are you sure you need the includePath setting for it to work?

**Comment by leonfs** [comment]
I'm running on a Mac M1 and I've been using the extension to fix a few IntelliSense problems (thanks by the way), but the whole discovery of files was really really slow. The loading of the symbols and suggestions would take ages. But, after reading the official docs of VSCode about C++ and seeing this issue thread, I decided to add the `includePath` property in the `c_cpp_properties.json`.  I have to say it worked like a treat! Now the whole discovery of files and IntelliSense is super fast. 

As a reference, my compiler path is: `"compilerPath": "/usr/bin/clang++"`

**Comment by boocs** [comment]
@leonfs 
I think that means you're using the Tag Parser Intellisense(which is fine). The latest UE4/UE5 use compile commands which override the includePath setting so includePath shouldn't be affecting anything. 

The compile commands must have include errors which forces VSCode to fall back to instead use Tag Parser Intellisense.

**Comment by leonfs** [comment]
You might be right, actually. How do you find those include errors @boocs? For some reason, navigating with IntelliSense the UE code is super fast, but when I have to navigate my project's code, it is super slow.

**Comment by boocs** [comment]
Tag Parser Intellisense is super fast but not as smart as the default Intellisense which is fast with cpp files but slow with header files because of how it works(updates too often whenever you type).

You should be able to look at the cpptools(C_CPP) extension logs. The setting is called Logging Level. Set it to debug.
I actually should just make a guide because it's really easy to check with the cpptools log. It'll take some time but let me do that.

**Comment by leonfs** [comment]
Also - In the logs of the C++ extension I see this:

```
For C++ source files, IntelliSenseMode was changed from "macos-clang-x64" to "macos-clang-arm64" 
```

**Comment by max-cis** [comment]
@leonfs just to make you aware that I am running an older i7 iMac so my compiler settings will differ from yours. Make sure you are not confusing the settings. There is another thread regarding M1 settings. 

@boocs I will have a look and report back. Sorry, got caught up in other things.

**Comment by boocs** [comment]
@leonfs I created a new issue for you called M1 Mac issues

**Comment by max-cis** [comment]
@boocs I started a brand new project and set the path in your extension. The settings are as follows: 
```
    "UEIntellisenseFixes.compiler.path": "/usr/bin/clang",
    "UEIntellisenseFixes.compiler.strictPath": true,
```

There were no immediate errors, but it was stuck on **Loading...** for ages. I then got an error with the include path and then it found things. It did take very long in comparison to when the includepath was set before. 

As far as I can tell, I am using the Default intelliSense. Is that not the proper one? The setting is set as follows:

```
"C_Cpp.autocomplete": "Default"
```
Any way to check if it is using the correct IntelliSense? Also, I haven't switched on your optimisation setting.

**Comment by boocs** [comment]
I don't know if you have have my latest extension version 3.2.0. But I changed the strict setting documentation to say only use it if the extension's path setting is the full compiler path. 

Make sure to change it to the full xcode/clang or clang++ path.

Makes me think I should just detect if it's a full path and error if it's not(when using the strict setting).

If you decide to remove the strict setting you should do a Generate Project File on your project.

**Comment by max-cis** [comment]
I did read that in your extension. However, when you read the output it states that it is better to actually have the box ticked. When you say full path you mean absolute rather than relative, right?

**Comment by boocs** [comment]
Yep, I guess my extension change wouldn't have worked if /user/bin/clang is also an absolute path on some systems. Maybe I'll just detect it on Mac and make sure if you use the strict setting that xcode is in the path somewhere.

So basically on Mac if using the strict setting your path setting should look something like this:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang or clang++ since clang++ is what ue5 default compiles with on Mac.

**Comment by boocs** [comment]
You can also remove the strict setting and Generate Project Files to get back the full xcode clang path in your compile commands file.

After you Generate Project Files, you can also check my extension log to see what the compile commands compiler path is being set to by the Unreal Engine.

```
Below is the compile commands first entry command. (WARNING) If getting errors and the compiler path is unexpected reset your UE project.
command: "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.32.31326\bin\HostX64\x64\cl.exe" @"D:\EJ\Documents\Unreal Projects\CPPTest\.vscode\compileCommands_CPPTest\CPPTest.211.rsp"
```
For example, my project's extension log above it says it's using "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.32.31326\bin\HostX64\x64\cl.exe"

**Comment by boocs** [comment]
Once you figure out the full path that Unreal chooses you can just copy and paste that into the path setting and reenable the strict setting.

**Comment by boocs** [comment]
Here's how I'll fix this.

- Recommend not using strict setting anymore. (Keep it for Windows users who want to compile with clang or future compiler bugs)
- Without the strict path setting it will default pull Unreal's clang path it uses in the compile commands and copy it to the extension's path setting. Now VSCode's compilerPath and compile commands setting will be in parity.

**Comment by boocs** [comment]
I released 3.3.0 with the path fixes. You can remove the strict setting and reset your project(Generate Project Files) to fix it.

The path setting will now to auto set based on what compiler Unreal chooses.

**Comment by boocs** [comment]
I've released my guide on how to check for VSCode config bugs. This is just a partial Include Errors release. Note: This will get reorganized and changed.

https://github.com/boocs/Unreal-VSCode-diy-config-check-github

**Comment by max-cis** [comment]
Ok, so I downloaded your latest version. Removed the strict flag and refreshed the project twice from Unreal 5. 
Restarted VSCode, got the extension to fix things and then I got errors with the includePath. See image below: 
<img width="1508" alt="Screenshot 2022-09-01 at 15 17 11" src="https://user-images.githubusercontent.com/47692294/187936978-a6412da9-2b04-42dd-95d0-df1164052004.png">

I checked c_cpp_properties.json and it is set up as follows: 
```
 {
            "name": "Test3Editor Editor Mac Development (Test3)",
            "intelliSenseMode": "clang-x64",
            "macFrameworkPath": [
                "/System/Library/Frameworks",
                "/Library/Frameworks"
            ],
            "compileCommands": "/Users/lmnicosia/CIS/Projects/UE5/Test3/.vscode/compileCommands_Test3.json",
            "cStandard": "c17",
            "cppStandard": "c++17",
            "compilerPath": "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++"
 }
```

I tried adding `"includePath": ["${workspaceFolder}/**"]` to c_cpp_properties.json but that didn't fix anything. 
I set the strict flag and that made the error on `Test3.h` go away but the other two include errors stayed. 

The current config in `settings.json` is: 

```
  "UEIntellisenseFixes.enableFixes": true,
    "UEIntellisenseFixes.cppStandard": "c++17",
    "C_Cpp.default.intelliSenseMode": "macos-clang-x64",
    "workbench.editor.untitled.hint": "hidden",
    "UEIntellisenseFixes.compiler.path": "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++",
    "C_Cpp.loggingLevel": "Debug",
    "UEIntellisenseFixes.compiler.strictPath": true,
```

Any ideas?

**Comment by max-cis** [comment]
Ok, so it appears that I managed to find a stable configuration. 

I removed the strict flag and left the compilerPath in your extension blank and restarted VSCode. 
I checked and it was auto-completed with: 
`
    "UEIntellisenseFixes.compiler.path": "/usr/bin/clang",
`
So far no errors. The output of your extension is as follows: 
```
Extension "UE Intellisense Fixes" 3.3.0 is now active!

Found Unreal Engine v5.0.3

Fixing compiler path in c_cpp_properties.json.
Auto updating extension's compiler setting to /usr/bin/clang
End fix c_cpp_properties compiler path.

Fixing compiler paths in compile commands.
Below is the compile commands first entry command. (WARNING) If getting errors and the compiler path is unexpected reset your UE project.
command: "/usr/bin/clang" @/Users/XXXXXX/CIS/Projects/UE5/Test3/.vscode/compileCommands_Test3/Test3.206.rsp
Compile Command Object wasn't modifed. Will not write file.
```

I have no idea why it doesn't like `"UEIntellisenseFixes.compiler.path": "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++"`.

Any ideas what's happening?

**Comment by boocs** [comment]
Seems like the project reset isn't working properly. It still has the /usr/bin/clang from the previous time.

Would it be possible to create a new project and see what compiler Unreal chooses for the compile commands (by checking extension log)?

(remember to level strict setting unchecked)

I'm also going to update my guide today for different ways to reset your project. Maybe one won't be broken. It'll be funny if it turns out your reset isn't broken and Unreal chooses usr/bin/clang for the compiler.

**Comment by max-cis** [comment]
I’ll quickly create a project and report back. Let’s see what’s happening!

**Comment by boocs** [comment]
Temporarily you can go into the path setting of the extension. It should now have to full xcode/clang++ path there.

Remove the ++ from the path and enable strict. 

Restart VSCode and see if the errors go away with the new project. Enabling strict copies the path setting to your compile commands file overwriting the compiler choice that Unreal chose.

**Comment by max-cis** [comment]
So, I did that and the errors keep on coming back. However, going to definition (F12) works! 
I checked XCode and I have 13.4.1. I am planning on deploying to iOS so I had to upgrade to whatever the latest version was. 

I wonder why usr/bin/clang works though.

**Comment by max-cis** [comment]
I am happy to help, however, I will try it tomorrow. 

While I have done lots of development experience in Windows, I am totally new to MacOS/iOS. Without your extension, I would have wasted a lot of time trying to figure all this out. Also, kind of keen on understanding how all of these tools are talking to each other!

**Comment by boocs** [comment]
> I wonder why usr/bin/clang works though.

Do you have more than one version of clang installed? Maybe that version is being used with usr/bin/clang and is more compatible with Intellisense. 

Unreal does use the C version of stdlib.h so maybe there's some clang or 'VSCode Intellisense' auto detection there which says you need to use clang and not clang++

**Comment by max-cis** [comment]
As far as I can tell I only have just the one. 
```
XXXX@XX:~$ clang --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
XXXX@XX:~$ clang++ --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

I could try to use the /usr/bin/clang++ and see how it goes...

**Comment by boocs** [comment]
In your new project does it have this like you specified above or did you add this yourself?:
"macFrameworkPath": [
                "/System/Library/Frameworks",
                "/Library/Frameworks"
            ],

Just wondering because Unreal shouldn't really be adding this line since they won't work with compile commands(unless I'm mistaken)

**Comment by max-cis** [comment]
I wonder if I need to set `C_Cpp.default.includePath` to make it work...

**Comment by boocs** [comment]
> I didn't add that, it just came out of unreal like that.

Strange. Maybe it's for the Tag Parser or something specific that VSCode needs for Macs.

Something else to try: 

I was reading about VSCode compilerPath setting and it says you can skip the compiler query by setting it to an empty string. This could rule out a VSCode query error.

Trouble is my extension won't allow you to set it to that without causing problems.

Something to try:

1. Remove the strict setting 
2. Create a new project and let my extension run (new project with the full xcode/c++ path)
3. Disable my extension for the new project.
4. In your test project open  .vscode/c_cpp_properties.json
5. Change both compilerPath settings to an empty string
6. Restart VSCode and see what happens

**Comment by max-cis** [comment]
Still broken! :( 
At this point, I am not sure if reverting back to `/usr/bin/clang` will make the errors go away!

**Comment by boocs** [comment]
> I wonder if I need to set `C_Cpp.default.includePath` to make it work...

includePaths are ignored because of the use of compile commands/response files

You can check out all your project's include paths by opening the response file listed in my extension logs:
```
Below is the compile commands first entry command. (WARNING) If getting errors and the compiler path is unexpected reset your UE project.
command: "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++" @/Users/XXXX/CIS/Projects/UE5/MyProject2/.vscode/compileCommands_MyProject2/MyProject2.204.rsp
```

The new guide, I mentioned, will also show the detected include with the use of the C++ extension logs.

**Comment by boocs** [comment]
Ah, never mind about that compilerPath being set to "". I tried it on Windows and my Intellisense failed as well.

I wonder how much a cheap mac is that could create UE5 projects? lol

**Comment by max-cis** [comment]
Cheap Macs! lol 
To be honest, not a fan. Just this project requires a Mac to deploy to iOS. 

I noticed that your extension doesn't seem to respect the strict flag. I tried reverting back to /usr/bin/clang and it wrote it over with the long path. However, after overwriting it manually with /usr/bin/clang and resetting the database the errors went away. The C++ extension output mentioned lots of tag parsing though. 

Anyway, I will try your guide tomorrow and see if I can extract more info. I'm off to have dinner! 
Ta

**Comment by boocs** [comment]
Yeah the tag parser can take 15+ minutes when it first parses your project's symbols. You have to wait for it to finish to use my guide. 

There is another command for, the VSCode cpptool extension, logs but I like the real time logs better (I've had the two different logs disagree before)

**Comment by boocs** [comment]
> I noticed that your extension doesn't seem to respect the strict flag.

That's just my lazy coding. When you enable the strict flag you have to restart VSCode for it to work ( I guess I should mention that somewhere).

**Comment by boocs** [comment]
I been thinking about that macFrameworkPath setting and thinking maybe this is the bug this entire time. I think for clang++ to work it needs the full absolute path to the system libraries found in the xcode path. Or maybe the relative paths just aren't working the full Xcode clang++ path.

I found this on stackoverflow:
```
"macFrameworkPath": [
                "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks"
            ],
```

Does your Xcode have a similar path like that?

So I would try:
- Remove strict setting
- Make sure project is using full Xcode clang++ path
- Add the full Xcode System/Library/Frameworks path to the "macFrameworkPath" setting
- Restart VSCode

Does it work now?

Also can you search "/Applications/Xcode.app/Contents/Developer/Platforms" for stdlib.h to see where it is located?
Or maybe just search your whole computer to see the different locations it's at.

**Comment by max-cis** [comment]
Ok, gave it a try. Still same errors: 

Full output of the C++ debug log here https://pastebin.com/67FeQPuC

The `c_cpp_properties.json` is as follows: 

```
{
    "configurations": [
        {
            "name": "MyProject2Editor Editor Mac Development (MyProject2)",
            "intelliSenseMode": "clang-x64",
            "macFrameworkPath": [
				"/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks",
                "/System/Library/Frameworks",
                "/Library/Frameworks"
            ],
            "compileCommands": "/Users/XXXX/CIS/Projects/UE5/MyProject2/.vscode/compileCommands_MyProject2.json",
            "cStandard": "c17",
            "cppStandard": "c++17",
            "compilerPath": "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++"
        },
        {
            "name": "Mac",
            "intelliSenseMode": "clang-x64",
            "macFrameworkPath": [
				"/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks",
                "/System/Library/Frameworks",
                "/Library/Frameworks"
            ],
            "compileCommands": "/Users/XXXX/CIS/Projects/UE5/MyProject2/.vscode/compileCommands_Default.json",
            "cStandard": "c17",
            "cppStandard": "c++17",
            "compilerPath": "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++"
        }
    ],
    "version": 4
}
```

and `settings.json` as follows: 

```
    "UEIntellisenseFixes.enableFixes": true,
    "UEIntellisenseFixes.cppStandard": "c++17",
    "C_Cpp.default.intelliSenseMode": "macos-clang-x64",
    "C_Cpp.loggingLevel": "Debug",
    "UEIntellisenseFixes.compiler.path": "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++",
```

I also used `find` to search for `stdlib.h` locations and found several. They are mostly related to different SDKs (didn't know I had more than one) and for different platforms: 

```
/Library/Developer/CommandLineTools/usr/include/c++/v1/stdlib.h
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/stdlib.h
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/stdlib.h
/System/Volumes/Data/Library/Developer/CommandLineTools/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/stdlib.h
/System/Volumes/Data/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/DriverKit.platform/Developer/SDKs/DriverKit.sdk/System/DriverKit/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/DriverKit.platform/Developer/SDKs/DriverKit.sdk/System/DriverKit/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS.sdk/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/WatchSimulator.platform/Developer/SDKs/WatchSimulator.sdk/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/WatchSimulator.platform/Developer/SDKs/WatchSimulator.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/stdlib.h
/System/Volumes/Data/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/stdlib.h
/System/Volumes/Data/opt/local/var/macports/sources/rsync.macports.org/macports/release/tarballs/base/vendor/tcl8.5.19/compat/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/DriverKit.platform/Developer/SDKs/DriverKit.sdk/System/DriverKit/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/DriverKit.platform/Developer/SDKs/DriverKit.sdk/System/DriverKit/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS.sdk/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS.sdk/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/WatchSimulator.platform/Developer/SDKs/WatchSimulator.sdk/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/WatchSimulator.platform/Developer/SDKs/WatchSimulator.sdk/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator.sdk/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/c++/v1/stdlib.h
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/include/stdlib.h
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/stdlib.h
/opt/local/var/macports/sources/rsync.macports.org/macports/release/tarballs/base/vendor/tcl8.5.19/compat/stdlib.h
```

Also, I followed your guide. The flame is on. 

I also tried adding to the macframeworks path /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++` instead of the one you suggested as that contains stdlib.h but still gave same errors.

I am surprised that it is not finding it!

**Comment by max-cis** [comment]
Sure, will do.
I think that one defaults to the tag parser though.

Best,

*Max Nicosia*
Director
Cambridge Intelligent Systems UK Ltd.

***@***.***

Salisbury House
Station Road
Cambridge
England
CB1 2LA

**Comment by max-cis** [comment]
Hey, can you message me your email. Apparently, the log is way longer when using /usr/bin/clang and it is beyond the limit for paste bin.

**Comment by boocs** [comment]
You can cut out all Unreal Paths if that makes it shorter. If it's working but only uses tag parser then that seems weird.

I was looking at Macs and Unreal Engine requirements. The only hard requirement is Using the latest Monterey OS. I saw that I can get a ~$200 late 2014 Mac Mini that can run that OS. I just need to be sure it will be able to create Unreal Projects in UE4/UE5. It doesn't even need to run the engine.

**Comment by max-cis** [comment]
Even though I removed all the Epic ones, it still was over pastebin's limit. So, I split it into two. 
I guess that could work. I don't know if you can install Monterey on such an old mac. You should check if they have a limit... 

https://pastebin.com/JuBgKDYh

https://pastebin.com/CiNqDCBP

**Comment by boocs** [comment]
There is something weird happening. Both logs say your clang gets detected as version 11

When I use clang 13.0.1 on Windows this is what is shown in logs:
```
other: --clang_version=130001
```

Will have to research and look at the logs more.

**Comment by max-cis** [comment]
That's pretty weird! Using `which` and `--version` yields the following: 

```
XXX@XXX:~$ which clang
/usr/bin/clang
XXX@XXX:~$ clang --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
XXX@XXX:~$
```

**Comment by boocs** [comment]
In your logs it also shows the clang version as 13.1.6 as well so I still think something is wrong with the mac system includes which is why VSCode detects it as version 11.

I was going to rent a cloud Mac to figure this out but if you want to keep going we can.

Here's what I want to try:
* Full xcode/clang++ path to 13.1.6 compile commands and compilerPath setting
* Remove both "macFrameworkPath" settings from your c_cpp_properties.
* Have c_cpp_properties.json open (this will prevent any Default intellisense logs)
* Restart VSCode
* Make sure both Intellisense icons aren't showing
* Restart VSCode again
* Post the log following this guide(I've updated it) https://github.com/boocs/Unreal-VSCode-diy-config-check-github

I want to see any mac paths that are included, if any, without the "macFrameworkPath" setting so we get a baseline.

**Comment by max-cis** [comment]
Happy to keep the helping, but today I am fighting another battle and I will be away on Sunday. Will try on Monday :)

**Comment by boocs** [comment]
No problem, I decided to rent the mac anyway because I need to test M1 Macs as well.

**Comment by boocs** [comment]
Don't bother doing it. The framework setting is something else and has nothing to do with includes. VSCode auto includes system paths for Mac just like it does on Windows. So it could be VSCode just isn't including the path to stdlib.h.

Still on going but thought I'd give an update.

**Comment by boocs** [comment]
So this must be a Mac C++ extension bug? 

I never even used my extension and just used /usr/bin/clang++ in the compile command file and it worked. Looking at the C++ extension logs shows it including a bunch of paths added from here:
```
Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/
```

Also usr/bin/clang works as well.

Even when it works it says clang_version=110000

So basically an easy fix for Unreal 5.0.3. I don't need to change anything on MacOS except the compiler path in the compile commands to /usr/bin/clang++

**Comment by max-cis** [comment]
That's interesting. When I first set up the VSCode in MacOS and used the compiler path it didn't work at all. That's why I went searching and came across your extension. 
So, does it work straight up if I set up the compiler path in the c_cpp extension to `usr/bin/clang++`?

**Comment by boocs** [comment]
Yeah that's because the compiler in the compile commands file overrides the compilerPath setting in c_cpp_properties.

So I I think I'm going to change my extension and have 2 compiler paths. One for c_cpp_properties.json where I have it as an error if it's set. Having it blank will remove the compilerPath setting from c_cpp_properties.

One for compile commands file, compiler path,  where I will default it to /usr/bin/clang++ for Macs. Or maybe I just won't auto set it and just explain in the instructions that they should use /usr/bin/clang++

**Comment by boocs** [comment]
I just kept it simple. One path setting no strict setting. If path isn't set it gets pulled from compile command compiler path and set.
Once set it will set both  compile commands compiler and compilerPath setting if they are different.

I tested it on my cloud M1 Mac and it works as long as I have the path set to /usr/bin/clang++

**Comment by max-cis** [comment]
Ok, cool! Have you created a new release? 

Also, I got two questions you may know about. I cannot find any info atm. 

1) When trying to build from VSCode the debug build, not the GameDebug I get the error: 
`ERROR: Targets cannot be built in the Debug configuration with this engine distribution.`
Any idea why? GameDebug does build fine. 

2) Is there any way to ensure that the Unreal Editor is not killed after you stop debugging from VSCode? As it takes a while to restart the Editor. If I have the Editor open, it just opens a copy for debugging. It doesn't attach to the running instance. 

Thanks!

**Comment by boocs** [comment]
Yep 3.4.0 is released!

> ...trying to build from VSCode the debug build

That's normal since you're not using the full source build that option wont work.

>  the Unreal Editor is not killed after you stop debugging

I don't think so. Sometimes when I'm just testing stuff I don't run debug. I just Build it using one of the VSCode Build Tasks. Once it's done Building successfully I just go to the project folder and double click the *.uproject file to run the project. You must build the 'Editor' suffix build if you want it to load into the Editor.

![image](https://user-images.githubusercontent.com/62588629/188949947-3c1c80b7-c5d3-4dc6-ba31-1367ecd8ef65.png)

That's the Build task I usually run on Windows. TestFps is my project name.

**Comment by max-cis** [comment]
Oh, ok, so I don't need to worry about the plain debug one. 

I guess if I need to inspect variables in VSCode the editor is always killed then. I was hoping I could save the time of restarting the editor each time I hit stop in the debugger. Oh well...

**Comment by max-cis** [comment]
Hey, so it's been a while. I just updated the extension and am getting errors on includes just like before. 

Settings are as follows: 
`
    "UEIntellisenseFixes.enableFixes": true,
    "UEIntellisenseFixes.cppStandard": "c++17",
    "C_Cpp.default.intelliSenseMode": "macos-clang-x64",
    "workbench.editor.untitled.hint": "hidden",
    "C_Cpp.loggingLevel": "Debug",
    "UEIntellisenseFixes.compiler.path": "/usr/bin/clang++",
`

Also, your extension is reporting some other problems. 

'
Found Unreal Engine v5.0.3


** Error **: 0: Couldn't create compileCommands from /Users/lmnicosia/CIS/Projects/UE5/Test3/.vscode/compileCommands_Default.json with key:MAIN! (not an error with UE 4.25.#)

** Error **: Did not create any compile commands with keyMAIN! (not an error with UE 4.25.#)

** Error **: Error in loadCompileCommandsFromWorkspace!

** Error **: This fixable project hasn't been initialized properly and is invalid.

** Error **: Couldn't create file watchers!

*** Number of error messages: 5
*** Number of warning messages: 0
If you get any errors you can try restarting VSCode to check if they've been fixed.

Extension is done.
'

I did update XCode recently... Maybe that broke everything... 
Any ideas?

**Comment by max-cis** [comment]
Indeed. I am in the process of downgrading. I bloody hate this walled garden of Eden!

**Comment by kristianmandrup** [comment]
> What XCode version? I know XCode 14/clang 14.0.0 can cause errors on Mac

Really? What errors? I'm on 14 as well.
